### PR TITLE
The labour of love

### LINKED ISSUE #30: Incorrect tagging of span:first-of-type
Hi!

When I have the following less:

```css
span:first-of-type {
  font-weight: bold;
}
```

Within another block of less it's incorrectly marked as `support type property-name css` but if I use any other html types or classes or even `span:last-of-type` it works as expected and applies `keyword control html elements` or some other slightly modified combination based on the value.

It just seems to be `span:first-of-type` that is the issue. I've had a look over the grammar but I don't see the issue, the rules all include first-of-type where applicable.

### LINKED ISSUE #37: Incorrect highlighting of hex colors on lines with variable interpolation
```less
a {
  background: #0F0F0F url("@{variable}/test.png");
}
```

This has incorrect highlighting on the hex color in Atom. The decimal numbers are correctly colored, but the hash and F digits are not. Taking out the `@{variable}` line fixes it.

Tested in Atom 1.0.2 on Windows.

### LINKED ISSUE #39: text-overflow is not highlighted
Previously filed at https://github.com/atom/language-css/issues/41, text-overflow is not highlighted.

### PR DESCRIPTION
**Pull request by braver**
- fix #39 
- fix #37
- fix #30
- add the `flex` property value
- add svg [elements](http://www.w3.org/TR/SVG11/eltindex.html)
- add svg [properties and values](http://www.w3.org/TR/SVG11/propidx.html) (only ones not already available for regular html elements)
- add the [currentColor](https://css-tricks.com/currentcolor/) value

--- 

*Update 30-10-2015 *

So, as I've found out that many of the issues of this grammar are related, it stopped making sense trying to tackle them one by one. It also doesn't make sense to make small, easy to oversee, PR's: it's tooo easy for one change to break something else, and it's impossible to make changes across multiple PR's and still be able to merge it.

I really want Atom to have the best every support for the LESS language, so that's going to take some work. This PR is devoted to that effort. Ambitions include:

- variable interpolation
- more built-in colours and font-names
- modern css stuff like `optimizeLegibility` and `background-blend-mode`
- more specs
- [font variant stuff](https://drafts.csswg.org/css-fonts-3/#descdef-font-variant)
- [ligatures](https://drafts.csswg.org/css-fonts-3/#propdef-font-variant-ligatures)

### CODE REVIEW

**Comment by braver** [comment]
I dunno, but halve the grammar doesn't even work:

before

I moved a bunch of stuff inside the property-value scope and that seems to clear a lot of stuff up. Note for instance the built-in font names.

after

I'm going to spend some more time on this, so consider it WIP until further notice.

**Comment by braver** [comment]
I'm not seeing any detrimental effects. @simurai could you perhaps test this PR a bit?

**Comment by simurai** [comment]
:+1:  Yup, got it checked out in dev mode.

Is it actually possible to "extend" a language? So that if I open a Less file, it would use the `language-css` and the `language-less` grammar. Then `language-less` could be simplified to only add the Less stuff on top of CSS. Or when making changes `language-css`, a Less file would get those changes too.

**Comment by 50Wliu** [comment]
I think `language-c` already uses that approach: https://github.com/atom/language-c/blob/0a21c625593b533de566797829294740525fe7da/grammars/c%2B%2B.cson#L26-L28

**Comment by braver** [comment]
That would be a great approach to keeping less and scss grammars lined up and minimal. I don't know what state the css grammar is in though, so I wouldn't directly assume it to be easy. 

I write Less every day, so I don't mind putting in some effort.

**Comment by simurai** [comment]
Probably hard to say which is more complete.. I was just wondering because I noticed that for example `currentColor` isn't highlighted in `language-css` either. So after the PR, it's probably behind.

Anyways, something to consider.. @50Wliu thanks for the pointer.

**Comment by braver** [comment]
Alright. My personal priority is on Less, so I'll focus on that first. It'll be interesting to then try to align the related languages (scss, sass, less, stylus, etc)... I'll keep it in mind for a rainy Sunday afternoon ;)

**Comment by burodepeper** [comment]
Hey @braver,
I've also been considering putting some effort towards simurai's idea of making css extendible. Let me know if you want to team up to tackle it.

PS, hallo mede-Delftenaar! : )

**Comment by braver** [comment]
I'm also definitely going to cover this in more specs, it's super hard to guard against regression otherwise with all these mega regexes. So unless anyone disagrees I'm going to go dig in on this one. Atom should have awesome support for its own core languages!

**Comment by 50Wliu** [comment]
> Atom should have awesome support for its own core languages!

+100!

**Comment by braver** [comment]
Err... I'm not done with this yet, but it's becoming a bit of an overhaul. Becoming more and more solid though.

Question: How do you feel about stuff like `background-blend-mode`? It's at "candidate recommendation" right now. Should we add stuff that isn't final like that? @simurai you seem to be on top of the bleeding edge and other more little talked about features in CSS, how do you feel about that and do you have like a list of favourites that aren't in this grammar yet?

**Comment by simurai** [comment]
> it's becoming a bit of an overhaul

Maybe needs a new title. ;)

> Question: How do you feel about stuff like `background-blend-mode`? It's at "candidate recommendation" right now. Should we add stuff that isn't final like that?

I think it's a good idea to be a bit forward looking and already add some of the coming stuff.. Maybe once 1-2 browser implemented a new standard.

Are you using some kind of reference.. like [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference) or [openweb](http://ref.openweb.io/CSS/) has everything nicely groupped.

**Comment by braver** [comment]
Yes, I'm mostly looking at mdn and the w3c specs pages. Still need to get more of the basics right, but I'll take a look at the bleeding edge stuff.

**Comment by braver** [comment]
Just as a little heads up, I'm scanning through all LESS files in Atom and in Bootstrap and recording the finals issues that need to be tackled. Getting there though.

**Comment by braver** [comment]
@simurai @50Wliu I'm basically done with this. All issues I could find when going through all LESS in Bootstrap and Atom have been fixed, except a bunch of vendor-prefix cases I might look into later. In comments in the PR I've added links to resources I used where possible. Specs have been added to make it easier to build on this. Open bugs except #24 (about go-to-declaration which I don't know anything about) and #34 (which I can't reproduce) have been fixed.

We should be able to share property-names, property-values, css functions, color-names and a bunch of other features between CSS, LESS, Stylus, etc. Most of them are named in the `'repository'`.

I'd love to get your feedback on this and hear if there is any area I can improve on before this is merged.

File: grammars/less.cson [hunk 1]
```diff
@@ -300,85 +327,183 @@
...
+    'include': '#gradient_builtin_function'
+  }
 ]
 'repository':
-  'property_list':
-    'begin': '\\{'
+  'variable_interpolation':
+    'match': '@{[a-zA-Z0-9_-]+}'
+    'name': 'variable.other.interpolation.less'
+  'string_double':
```
Comment (50Wliu) [inline grammars/less.cson]: I would merge the `string_double` and `string_single` patterns into a single one called `strings` since they're always included together.

File: grammars/less.cson [hunk 2]
```diff
@@ -116,7 +88,7 @@
     ]
   }
   {
-    'match': '(#)([0-9a-fA-F]{3}|[0-9a-fA-F]{6})\\b(?!.*?\\{)'
+    'match': '(#)([0-9a-fA-F]{3}|[0-9a-fA-F]{6})\\b(?!.*?(?<!@){)'
```
Comment (50Wliu) [inline grammars/less.cson]: What did this fix?

File: grammars/less.cson [hunk 3]
```diff
@@ -183,13 +141,35 @@
...
+        'name': 'keyword.control.at-rule.import.less'
       '2':
-        'name': 'punctuation.definition.keyword.css'
-    'match': '^\\s*((@)import\\b)'
+        'name': 'punctuation.definition.keyword.less'
+    'end': ';'
+    'endCaptures':
+      '0':
+        'name': 'punctuation.terminator.rule.css'
     'name': 'meta.at-rule.import.css'
```
Comment (50Wliu) [inline grammars/less.cson]: This will target the entire rule as `meta.at-rule.import.css`, including any leading whitespace due to the begin rule.

File: grammars/less.cson [hunk 4]
```diff
@@ -183,13 +141,35 @@
...
+    'patterns': [
+      {
+        'match': '(?<=\\\(|,|\\s)\\b(reference|inline|less|once|multiple|optional)\\b(?=\\\)|,)'
```
Comment (50Wliu) [inline grammars/less.cson]: I think the triple backslashes is a typos here.

File: grammars/less.cson [hunk 5]
```diff
@@ -251,11 +232,48 @@
     'name': 'variable.other.less'
   }
   {
-    "match": "\\!\\s*important",
+    'include': '#variable_interpolation'
+  }
+  (
+    'begin': '\\{'
```
Comment (50Wliu) [inline grammars/less.cson]: Curly brackets don't need to be escaped

File: grammars/less.cson [hunk 6]
```diff
@@ -278,15 +306,14 @@
...
   }
   {
     'match': '\\{|\\}'
-    'name': 'meta.brace.curly.less'
+    'name': 'meta.brace.curly.css'
   }
   {
-    'match': '\\(|\\)'
-    'name': 'meta.brace.round.less'
+    'include': '#brace_round'
```
Comment (50Wliu) [inline grammars/less.cson]: Why was this and not the curly braces one made into a repository?

File: grammars/less.cson [hunk 7]
```diff
@@ -300,85 +327,183 @@
     'match': ':'
     'name': 'punctuation.separator.key-value.css'
   }
+  {
+    'match': '\\btrue\\b'
```
Comment (50Wliu) [inline grammars/less.cson]: Does `false` not exist in less?

File: grammars/less.cson [hunk 8]
```diff
@@ -300,85 +327,183 @@
...
+    'name': 'constant.language.boolean.less'
+  }
+  {
+    'match': '\\bdefault\\b'
+    'name': 'support.function.default.less'
+  }
+  {
+    'match': '\\b(iscolor|isnumber|isstring|iskeyword|isurl)\\b'
```
Comment (50Wliu) [inline grammars/less.cson]: alphabetize

File: grammars/less.cson [hunk 9]
```diff
@@ -300,85 +327,183 @@
...
+    'name': 'support.function.type-checking.less'
+  }
+  {
+    'match': '\\b(ispixel|ispercentage|isem|isunit)\\b'
```
Comment (50Wliu) [inline grammars/less.cson]: ditto

File: grammars/less.cson [hunk 10]
```diff
@@ -300,85 +327,183 @@
...
+  'property_name_svg':
+    'captures':
+      '1':
+        'name': 'support.type.property-name.svg.css'
+    'match': '(?<![-a-z])(writing-mode|text-anchor|stroke-width|stroke-opacity|stroke-miterlimit|stroke-linejoin|stroke-linecap|stroke-dashoffset|stroke-dasharray|stroke|stop-opacity|stop-color|shape-rendering|marker-start|marker-mid|marker-end|lighting-color	|kerning|image-rendering|glyph-orientation-vertical|glyph-orientation-horizontal|flood-opacity|flood-color|fill-rule|fill-opacity|fill|enable-background|color-rendering|color-interpolation-filters|color-interpolation|clip-rule|clip-path)(?=\\s*:)'
```
Comment (50Wliu) [inline grammars/less.cson]: Extra whitespace after `lighting_color`

File: grammars/less.cson [hunk 11]
```diff
@@ -300,85 +327,183 @@
...
   'property_value':
-    'begin': '(?<!&)(:)\\s*(?!(\\s*{))'
+    'begin': '(?<!&)(:)\\s*(?!(\\s*{))(?!.*(?<!@){)'
```
Comment (50Wliu) [inline grammars/less.cson]: What did this fix?

File: grammars/less.cson [hunk 12]
```diff
@@ -300,85 +327,183 @@
...
+  'property_name_svg':
+    'captures':
+      '1':
+        'name': 'support.type.property-name.svg.css'
+    'match': '(?<![-a-z])(writing-mode|text-anchor|stroke-width|stroke-opacity|stroke-miterlimit|stroke-linejoin|stroke-linecap|stroke-dashoffset|stroke-dasharray|stroke|stop-opacity|stop-color|shape-rendering|marker-start|marker-mid|marker-end|lighting-color	|kerning|image-rendering|glyph-orientation-vertical|glyph-orientation-horizontal|flood-opacity|flood-color|fill-rule|fill-opacity|fill|enable-background|color-rendering|color-interpolation-filters|color-interpolation|clip-rule|clip-path)(?=\\s*:)'
   'property_value':
```
Comment (50Wliu) [inline grammars/less.cson]: property_value -> property_values?

File: grammars/less.cson [hunk 13]
```diff
@@ -300,85 +327,183 @@
...
+    'match': '\\b(abs|acos|alpha|argb|asin|atan|average|blue|calc|ceil|color|contrast|convert|convert|cos|darken|data-uri|desaturate|difference|e|escape|exclusion|extract|fade|fadein|fadeout|floor|format|green|greyscale|hardlight|hsl|hsla|hsv|hsva|hsvhue|hsvsaturation|hsvvalue|hue|length|lighten|lightness|luma|max|min|mix|mod|multiply|negation|overlay|percentage|pi|pow|red|replace|round|saturate|saturation|screen|sin|softlight|spin|sqrt|tan|unit)\\b'
     'name': 'support.function.any-method.builtin.less'
   'css_builtin_function':
-    'match': '\\b(rgb|rgba|hsl|hsla|url|scale|scale[XY]|translate|translate[XY]|rect|rotate|skew|skew[XY]|cubic-bezier|translate3d|blur)\\b'
+    'match': '\\b(url|translate3d|translate[XYZ]|translate|skew[XY]|skew|scale[XYZ]|scale|rotate3d|rotate[XYZ]|rotate|rgba|rgb|repeating-radial-gradient|repeating-linear-gradient|rect|radial-gradient|matrix3d|matrix|linear-gradient|hsla|hsl|drop-shadow|cubic-bezier|blur)\\b'
     'name': 'support.function.any-method.builtin.css'
   'gradient_builtin_function':
```
Comment (50Wliu) [inline grammars/less.cson]: Should the above 3 all be pluralized (functions)?

File: grammars/less.cson [hunk 14]
```diff
@@ -300,85 +327,183 @@
...
+    'captures':
+      '1':
+        'name': 'punctuation.definition.entity.css'
+    'match': '(:)(active|checked|default|dir|disabled|empty|enabled|extend|first-child|first-of-type|first|fullscreen|focus|hover|indeterminate|in-range|invalid|lang|last-child|last-of-type|left|link|not|nth-child|nth-last-child|nth-last-of-type|nth-of-type|only-child|only-of-type|optional|out-of-range|read-only|read-write|required|right|root|scope|shadow|target|valid|visited)(?!\\s*;)'
+    'name': 'entity.other.attribute-name.pseudo-class.css'
   'font_name':
-    'match': '(\\b(?i:arial|century|comic|courier|futura|garamond|georgia|helvetica|impact|lucida|symbol|system|tahoma|times|trebuchet|utopia|verdana|webdings|sans-serif|serif|monospace)\\b)'
+    'match': '(\\b(?i:arial|century|comic|courier|cursive|fantasy|futura|garamond|georgia|helvetica|impact|lucida|monospace|symbol|system|tahoma|times|trebuchet|utopia|verdana|webdings|sans-serif|serif)\\b)'
     'name': 'support.constant.font-name.css'
   'color_name':
```
Comment (50Wliu) [inline grammars/less.cson]: Pluralize this and font_name?

File: spec/grammar-spec.coffee [hunk 15]
```diff
@@ -64,6 +69,17 @@ describe "less grammar", ->
     expect(tokens[7].scopes).not.toContain 'meta.property-value.css'
     expect(tokens[11].scopes).not.toContain 'meta.property-value.css'
 
+  it "parses @media features", ->
```
Comment (50Wliu) [inline spec/grammar-spec.coffee]: Might be beneficial to test for the tokens of `@media` as well.

File: spec/grammar-spec.coffee [hunk 16]
```diff
@@ -95,27 +111,51 @@ describe "less grammar", ->
     expect(tokens[0]).toEqual value: "#", scopes: ['source.css.less', 'meta.selector.css', 'entity.other.attribute-name.id', 'punctuation.definition.entity.css']
     expect(tokens[1]).toEqual value: "abc-123", scopes: ['source.css.less', 'meta.selector.css', 'entity.other.attribute-name.id']
 
+  it "parses pseudo classes", ->
+    {tokens} = grammar.tokenizeLine(".foo:hover { span:last-of-type { font-weight: bold; } }")
+    expect(tokens[0]).toEqual value: ".", scopes: ['source.css.less', 'entity.other.attribute-name.class.css', 'punctuation.definition.entity.css']
+    expect(tokens[1]).toEqual value: "foo", scopes: ['source.css.less', 'entity.other.attribute-name.class.css']
+    expect(tokens[2]).toEqual value: ":", scopes: ['source.css.less', 'entity.other.attribute-name.pseudo-class.css', 'punctuation.definition.entity.css']
+    expect(tokens[3]).toEqual value: "hover", scopes: ['source.css.less', 'entity.other.attribute-name.pseudo-class.css' ]
+    expect(tokens[4]).toEqual value: " ", scopes: ['source.css.less']
```
Comment (50Wliu) [inline spec/grammar-spec.coffee]: No need to test the whitespace :P

File: spec/grammar-spec.coffee [hunk 17]
```diff
@@ -95,27 +111,51 @@ describe "less grammar", ->
...
-    expect(tokens[8]).toEqual value: "none", scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-value.css', 'support.constant.property-value.css']
+    expect(tokens[8]).toEqual value: "table-row", scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-value.css', 'support.constant.property-value.css']
     expect(tokens[9]).toEqual value: ";", scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-value.css', 'punctuation.terminator.rule.css']
     expect(tokens[10]).toEqual value: " ", scopes: ['source.css.less', 'meta.property-list.css']
     expect(tokens[11]).toEqual value: "}", scopes: ['source.css.less', 'meta.property-list.css', 'punctuation.section.property-list.end.css']
 
+  it 'parses font lists', ->
+    {tokens} = grammar.tokenizeLine '.foo { font-family: "Some Font Name", serif; }'
+    expect(tokens[5]).toEqual value: 'font-family', scopes: ['source.css.less', 'meta.property-list.css', 'support.type.property-name.css']
+    expect(tokens[9]).toEqual value: 'Some Font Name', scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-value.css', 'string.quoted.double.css']
```
Comment (50Wliu) [inline spec/grammar-spec.coffee]: Would absolutely love it if `font-family:` could be turned into a begin/end rule that matches everything following it as a font, but maybe that can come later :).

File: spec/grammar-spec.coffee [hunk 18]
```diff
@@ -145,3 +185,122 @@ describe "less grammar", ->
     expect(tokens[9]).toEqual value: ";", scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-value.css', 'punctuation.terminator.rule.css']
     expect(tokens[10]).toEqual value: " ", scopes: ['source.css.less', 'meta.property-list.css']
     expect(tokens[11]).toEqual value: "}", scopes: ['source.css.less', 'meta.property-list.css', 'punctuation.section.property-list.end.css']
+
+  it 'parses variable interpolation in selectors', ->
+    {tokens} = grammar.tokenizeLine '.@{selector} { color: #0ee; }'
```
Comment (50Wliu) [inline spec/grammar-spec.coffee]: Is the `.` relevant here?  If so, please test for it.

File: spec/grammar-spec.coffee [hunk 19]
```diff
@@ -145,3 +185,122 @@ describe "less grammar", ->
...
+    expect(tokens[1]).toEqual value: "foo", scopes: ['source.css.less', 'entity.other.attribute-name.class.css']
+    expect(tokens[2]).toEqual value: " ", scopes: ['source.css.less']
+    expect(tokens[3]).toEqual value: "{", scopes: ['source.css.less', 'meta.property-list.css', 'punctuation.section.property-list.begin.css']
+    expect(tokens[4]).toEqual value: " ", scopes: ['source.css.less', 'meta.property-list.css']
+    expect(tokens[5]).toEqual value: '@{property}', scopes: ['source.css.less', 'meta.property-list.css', 'variable.other.interpolation.less']
+    expect(tokens[6]).toEqual value: ":", scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-value.css', 'punctuation.separator.key-value.css']
+    expect(tokens[7]).toEqual value: " ", scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-value.css']
+
+  it 'parses options in import statements', ->
+    {tokens} = grammar.tokenizeLine '@import (optional, reference) "theme";'
```
Comment (50Wliu) [inline spec/grammar-spec.coffee]: Test for the comma please

File: spec/grammar-spec.coffee [hunk 20]
```diff
@@ -145,3 +185,122 @@ describe "less grammar", ->
...
+    expect(tokens[0]).toEqual value: "@", scopes: ['source.css.less', 'meta.at-rule.import.css', 'keyword.control.at-rule.import.less', 'punctuation.definition.keyword.less']
+    expect(tokens[1]).toEqual value: "import", scopes: ['source.css.less', 'meta.at-rule.import.css', 'keyword.control.at-rule.import.less']
+    expect(tokens[4]).toEqual value: "optional", scopes: ['source.css.less', 'meta.at-rule.import.css', 'keyword.control.import.option.less']
+    expect(tokens[7]).toEqual value: "reference", scopes: ['source.css.less', 'meta.at-rule.import.css', 'keyword.control.import.option.less']
+    expect(tokens[11]).toEqual value: "theme", scopes: ['source.css.less', 'meta.at-rule.import.css', 'string.quoted.double.css']
+
+  it 'parses built-in functions in property values', ->
+    {tokens} = grammar.tokenizeLine '.foo { border: 1px solid rgba(0,0,0); }'
```
Comment (50Wliu) [inline spec/grammar-spec.coffee]: Test for commas

File: spec/grammar-spec.coffee [hunk 21]
```diff
@@ -145,3 +185,122 @@ describe "less grammar", ->
...
+    expect(tokens[9]).toEqual value: "px", scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-value.css', 'keyword.other.unit.css']
+    expect(tokens[11]).toEqual value: "solid", scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-value.css', 'support.constant.property-value.css']
+    expect(tokens[13]).toEqual value: "rgba", scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-value.css', 'support.function.any-method.builtin.css']
+    expect(tokens[14]).toEqual value: "(", scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-value.css', 'meta.brace.round.css']
+    expect(tokens[15]).toEqual value: "0", scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-value.css', 'constant.numeric.css']
+    expect(tokens[21]).toEqual value: ";", scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-value.css', 'punctuation.terminator.rule.css']
+
+  it 'parses linear-gradient', ->
+    {tokens} = grammar.tokenizeLine '.foo { background: linear-gradient(white, black); }'
```
Comment (50Wliu) [inline spec/grammar-spec.coffee]: Commas

File: spec/grammar-spec.coffee [hunk 22]
```diff
@@ -145,3 +185,122 @@ describe "less grammar", ->
...
+
+  it 'parses blend modes', ->
+    {tokens} = grammar.tokenizeLine '.foo { background-blend-mode: color-dodge; }'
+    expect(tokens[5]).toEqual value: "background-blend-mode", scopes: ['source.css.less', 'meta.property-list.css', 'support.type.property-name.css']
+    expect(tokens[6]).toEqual value: ":", scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-value.css', 'punctuation.separator.key-value.css']
+    expect(tokens[7]).toEqual value: " ", scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-value.css']
+    expect(tokens[8]).toEqual value: "color-dodge", scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-value.css', 'support.constant.property-value.css']
+    expect(tokens[9]).toEqual value: ";", scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-value.css', 'punctuation.terminator.rule.css']
+
+  it 'parses nested multiple lines with pseudo-classes', ->
```
Comment (50Wliu) [inline spec/grammar-spec.coffee]: Shouldn't this test be with the other pseudo-classes one?

File: spec/grammar-spec.coffee [hunk 23]
```diff
@@ -145,3 +185,122 @@ describe "less grammar", ->
...
+    expect(lines[0][7]).toEqual value: ',', scopes: ['source.css.less', 'meta.property-list.css', 'punctuation.separator.list.css']
+    expect(lines[1][0]).toEqual value: 'p', scopes: ['source.css.less', 'meta.property-list.css', 'keyword.control.html.elements' ]
+    expect(lines[1][1]).toEqual value: ':', scopes: ['source.css.less', 'meta.property-list.css', 'entity.other.attribute-name.pseudo-class.css', 'punctuation.definition.entity.css']
+    expect(lines[1][2]).toEqual value: 'active', scopes: ['source.css.less', 'meta.property-list.css', 'entity.other.attribute-name.pseudo-class.css']
+    expect(lines[1][3]).toEqual value: ' ', scopes: ['source.css.less', 'meta.property-list.css']
+    expect(lines[1][4]).toEqual value: '{', scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-list.css', 'punctuation.section.property-list.begin.css']
+    expect(lines[1][5]).toEqual value: ' ', scopes: ['source.css.less', 'meta.property-list.css', 'meta.property-list.css']
+
+  it 'parses variable interpolation in imports', ->
```
Comment (50Wliu) [inline spec/grammar-spec.coffee]: Ditto regarding this one and interpolation/imports

File: spec/grammar-spec.coffee [hunk 24]
```diff
@@ -145,3 +185,122 @@ describe "less grammar", ->
...
+    {tokens} = grammar.tokenizeLine '@import "@{var}/tidal-wave.less";'
+    expect(tokens[0]).toEqual value: "@", scopes: ['source.css.less', 'meta.at-rule.import.css', 'keyword.control.at-rule.import.less', 'punctuation.definition.keyword.less']
+    expect(tokens[1]).toEqual value: "import", scopes: ['source.css.less', 'meta.at-rule.import.css', 'keyword.control.at-rule.import.less']
+    expect(tokens[2]).toEqual value: " ", scopes: ['source.css.less', 'meta.at-rule.import.css']
+    expect(tokens[3]).toEqual value: "\"", scopes: ['source.css.less', 'meta.at-rule.import.css', 'string.quoted.double.css', 'punctuation.definition.string.begin.css']
+    expect(tokens[4]).toEqual value: "@{var}", scopes: ['source.css.less', 'meta.at-rule.import.css', 'string.quoted.double.css', 'variable.other.interpolation.less']
+
+  it 'parses variable interpolation in urls', ->
```
Comment (50Wliu) [inline spec/grammar-spec.coffee]: Ditto

**Comment by 50Wliu** [comment]
Looking pretty good!

**Comment by 50Wliu** [comment]
In addition: what's the rule for whether something should get tagged as `.less` or `.css`?

**Comment by braver** [comment]
@50Wliu thanks for the feedback, I'll look into it. You're seeing a lot of inconsistencies that were already there, but I think it's a good time to weed that out.

I tried tagging stuff as .less only if it's .less specific, like the built in functions for darken/lighten. What should happen though is everything css like the properties and color names should be taken out and reused from the css language, so everything in this package can get tagged less. I'll do a sanity check, but I don't think it will be perfect until we reuse across the languages.

File: grammars/less.cson (hunk 13 above)
Comment (braver) [inline grammars/less.cson]: You mean the label (line 505)?

File: grammars/less.cson (hunk 1 above)
Comment (braver) [inline grammars/less.cson]: done: c98bd88

File: grammars/less.cson (hunk 3 above)
Comment (braver) [inline grammars/less.cson]: I thought it might be a good idea to scope the entire statement as `meta.at-rule.import.css`, so we can limit what kind of scopes are allowed inside that statement. @50Wliu are you opposed to that, or just the inclusion of leading whitespace?

Comment (braver) [inline grammars/less.cson]: addressed in 19db7a4 keeping the statement wrapped

File: grammars/less.cson (hunk 4 above)
Comment (braver) [inline grammars/less.cson]: addressed in 266435a and d9cb17f

File: grammars/less.cson (hunk 5 above)
Comment (braver) [inline grammars/less.cson]: addressed in 20228c7

File: grammars/less.cson (hunk 6 above)
Comment (braver) [inline grammars/less.cson]: This is the only one that is used more that once. I'm on the fence about this. Moving the other braces to the repository as well would be consistent, but it also adds code. If you have a clear preference I don't mind changing it.

File: grammars/less.cson (hunk 7 above)
Comment (braver) [inline grammars/less.cson]: It doesn't look like it. Basically, only `true` is truthy, everything else is false: http://lesscss.org/features/#mixin-guards-feature-guard-comparison-operators

File: grammars/less.cson (hunk 8 above)
Comment (braver) [inline grammars/less.cson]: done: ef4c6bd

File: grammars/less.cson (hunk 9 above)
Comment (braver) [inline grammars/less.cson]: done: ef4c6bd

File: grammars/less.cson (hunk 10 above)
Comment (braver) [inline grammars/less.cson]: I fixed the whitespace, but I'm not sure about the `-`. What should the effect be? There aren't any specs for it and I'm not sure what it should accomplish in the first place. It was like that so I kept it.

File: grammars/less.cson (hunk 12 above)
Comment (braver) [inline grammars/less.cson]: done: 892ca1e

File: grammars/less.cson (hunk 13 above)
Comment (braver) [inline grammars/less.cson]: I pluralised everything in the repository: 892ca1e

File: spec/grammar-spec.coffee (hunk 15 above)
Comment (braver) [inline spec/grammar-spec.coffee]: done: 905ef1c

File: spec/grammar-spec.coffee (hunk 16 above)
Comment (braver) [inline spec/grammar-spec.coffee]: Well, sometimes it's useful to check the scoped of the whitespace as well. In this case it might be a bit much, but it does guard against accidentally including whitespace where a lookahead would be better. So, I'm kind of in favour of keeping this and testing too much, over hunting down the spec for things we might not need to test.

File: spec/grammar-spec.coffee (hunk 17 above)
Comment (braver) [inline spec/grammar-spec.coffee]: Yeah, why not open an 'enhancement' issue for that ;)

File: spec/grammar-spec.coffee (hunk 18 above)
Comment (braver) [inline spec/grammar-spec.coffee]: It actually isn't really, it wasn't being tokenised at all. In 42c4632 I changed it so the variable interpolation is part of the `attribute-name.class` and `attribute-name.id` scope, which also means the `.` is correctly tokenised. Specs updated to match.

File: spec/grammar-spec.coffee (hunk 20 above)
Comment (braver) [inline spec/grammar-spec.coffee]: Isn't this the same as above? 7cd7a00

File: spec/grammar-spec.coffee (hunk 19 above)
Comment (braver) [inline spec/grammar-spec.coffee]: done: 7cd7a00

File: spec/grammar-spec.coffee (hunk 21 above)
Comment (braver) [inline spec/grammar-spec.coffee]: done:  000f07d

File: spec/grammar-spec.coffee (hunk 22 above)
Comment (braver) [inline spec/grammar-spec.coffee]: done: 8b1df91

File: spec/grammar-spec.coffee (hunk 23 above)
Comment (braver) [inline spec/grammar-spec.coffee]: done: 2f753f9

File: spec/grammar-spec.coffee (hunk 24 above)
Comment (braver) [inline spec/grammar-spec.coffee]: done: 2f753f9

File: grammars/less.cson (hunk 2 above)
Comment (braver) [inline grammars/less.cson]: This fixes #37. I'll add a test for it.

**Comment by braver** [comment]
The `foldingStartMarker` and `foldingStopMarker` don't seem to actually do anything in Atom. I think they're remnants from Textmate. Should I take them out?

**Comment by braver** [comment]
I've added some inline comments. All changes to lookahead are now also covered by specs.

File: grammars/less.cson (hunk 3 above)
Comment (50Wliu) [inline grammars/less.cson]: Just the whitespace.  I'll take a look at 19db7a4.

**Comment by 50Wliu** [comment]
> The foldingStartMarker and foldingStopMarker don't seem to actually do anything in Atom. I think they're remnants from Textmate. Should I take them out?

Not sure.  Currently Atom handles folding through indentation but that could change later on and then these regex patterns may be useful.

**Comment by 50Wliu** [comment]
Looks good to me, but I'll defer the final decision to @simurai.

**Comment by braver** [comment]
Well if that's the case, I'd say we can always pull it out of the history. No sense in keeping non-functioning code around "just in case".

**Comment by 50Wliu** [comment]
Sure, then why not.

**Comment by braver** [comment]
@simurai Anything you want me to look at, or can you otherwise greenlight this?

**Comment by simurai** [comment]
Sorry, meant to reply earlier. Looks good to me too.

When testing "every" [CSS properties](https://gist.github.com/simurai/40902209f695b2d080e1), there are a few missing. But compared to before or `language-css` it's a huge improvement. :clap:

ok, let's :ship: this.

**Comment by simurai** [comment]
@braver thanks for the :heart:.

**Comment by braver** [comment]
Awesome, thanks guys :+1:

**Comment by braver** [comment]
@simurai since you've got a full list of property names, we can simply check the list in this grammar against it... Perhaps I'll open a pr for this later ;)

@burodepeper if you want to look into making css extensible, you should be able to lift stuff like property names, color names, etc. from this grammar. Some of it is affected by things like Less stealing the calc function, or by variable interpolation, but most of the simple list matching should be transferable between css and preprocessors.
